build-logic-commons/build-scan/src/main/kotlin/gradlebuild.cache-miss-monitor.gradle.kts

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import com.gradle.scan.plugin.BuildScanExtension
+import java.util.concurrent.atomic.AtomicBoolean
+
+val cacheMissTagged = AtomicBoolean(false)
+val buildScan = extensions.findByType<BuildScanExtension>()
+
+if (System.getenv("TEAMCITY_VERSION") != null) {
+    gradle.taskGraph.afterTask {
+        if (buildCacheEnabled() && isCacheMiss() && isNotTaggedYet()) {
+            buildScan?.tag("CACHE_MISS")
+        }
+    }
+}
+
+fun buildCacheEnabled() = gradle.startParameter.isBuildCacheEnabled
+
+fun isNotTaggedYet() = cacheMissTagged.compareAndSet(false, true)
+
+fun Task.isCacheMiss() = !state.skipped && state.failure == null && (isCompileCacheMiss() || isAsciidoctorCacheMiss())
+
+fun Task.isCompileCacheMiss() = isMonitoredCompileTask() && !isExpectedCompileCacheMiss()
+
+fun isAsciidoctorCacheMiss() = isMonitoredAsciidoctorTask() && !isExpectedAsciidoctorCacheMiss()
+
+fun Task.isMonitoredCompileTask() = (this is AbstractCompile || this.isClasspathManifest()) && !this.isKotlinJsIrLink()
+
+fun Task.isClasspathManifest() = this.javaClass.simpleName.startsWith("ClasspathManifest")
+
+// https://youtrack.jetbrains.com/issue/KT-49915
+fun Task.isKotlinJsIrLink() = this.javaClass.simpleName.startsWith("KotlinJsIrLink")
+
+fun isMonitoredAsciidoctorTask() = false // No asciidoctor tasks are cacheable for now
```

Comment:
It seems to me that build logic commons shouldn't know about asciidoctor at all. Can we maybe make this plugin somehow configurable, so that the plugin in `build-logic/profiling` can add some more exclusions?

We can also leave it for now in build logic commons, I suppose it is not too horrible.

Reply:
Yeah... I would like not to introduce more complexity for now...